src/sat/sat_clause.cpp

```diff
@@ -123,53 +123,80 @@ namespace sat {
 
     clause_allocator::clause_allocator():
         m_allocator("clause-allocator") {
-#ifdef _AMD64_
+#if defined(_AMD64_) 
         m_num_segments = 0;
 #endif
     }
 
     clause * clause_allocator::get_clause(clause_offset cls_off) const {
-#ifdef _AMD64_
-        return reinterpret_cast<clause *>(m_segments[cls_off & c_aligment_mask] + (static_cast<size_t>(cls_off) & ~c_aligment_mask));
+#if defined(_AMD64_) 
+#if defined (Z3DEBUG)
+        clause const* result;
+        if (((cls_off & c_alignment_mask) == c_last_segment)) { 
+            unsigned id = cls_off >> c_cls_alignment;
+            bool check = m_last_seg_id2cls.find(id, result);
+            SASSERT(check);
+            return const_cast<clause*>(result);
+        }
+#endif
+        return reinterpret_cast<clause *>(m_segments[cls_off & c_alignment_mask] + (static_cast<size_t>(cls_off) & ~c_alignment_mask));
 #else
         return reinterpret_cast<clause *>(cls_off);
 #endif
     }
 
-#ifdef _AMD64_
-    unsigned clause_allocator::get_segment(size_t ptr) {
-        SASSERT((ptr & c_aligment_mask) == 0);
-        ptr &= ~0xFFFFFFFFull; // Keep only high part
+#if defined(_AMD64_) 
+    unsigned clause_allocator::get_segment(clause const* cls) {
+        size_t ptr = reinterpret_cast<size_t>(cls);
+
+        SASSERT((ptr & c_alignment_mask) == 0);
+        ptr &= 0xFFFFFFFF00000000ull; // Keep only high part
         unsigned i = 0;
         for (i = 0; i < m_num_segments; ++i)
-            if (m_segments[i] == ptr)
-                return i;
+          if (m_segments[i] == ptr) 
+            return i;
         i = m_num_segments;
-        m_num_segments++;
-        SASSERT(m_num_segments <= c_max_segments);
-        if (i >= c_max_segments)
+#if defined(Z3DEBUG)
+        SASSERT(i <= c_last_segment);
```

Comment:
The `SASSERT(i <= c_last_segment)` is the same for both `#if .. #else` blocks so I'd recommend hoisting it out to avoid code duplication and for clarity.